src/Compiler/Microsoft.AspNetCore.Razor.Language/src/RazorDiagnostic.cs

```diff
@@ -1,63 +1,93 @@
+        => new(descriptor, span: null, args: null);
 
-    public abstract string GetMessage(IFormatProvider formatProvider);
+    public static RazorDiagnostic Create(RazorDiagnosticDescriptor descriptor, SourceSpan? span)
+        => new(descriptor, span, args: null);
 
-    public string GetMessage() => GetMessage(null);
+    public static RazorDiagnostic Create(RazorDiagnosticDescriptor descriptor, params object[] args)
+        => new(descriptor, span: null, args);
 
-    public abstract bool Equals(RazorDiagnostic other);
+    public static RazorDiagnostic Create(RazorDiagnosticDescriptor descriptor, SourceSpan? span, params object[] args)
+        => new(descriptor, span, args);
 
-    public abstract override int GetHashCode();
+    public string GetMessage()
+        => GetMessage(formatProvider: null);
 
-    public static RazorDiagnostic Create(RazorDiagnosticDescriptor descriptor, SourceSpan span)
+    public string GetMessage(IFormatProvider? formatProvider)
+        => string.Format(formatProvider, Descriptor.MessageFormat, Args);
+
+    public override bool Equals(object? obj)
+        => obj is RazorDiagnostic diagnostic
+            ? Equals(diagnostic)
+            : false;
+
+    public bool Equals(RazorDiagnostic other)
     {
-        if (descriptor == null)
+        if (!Descriptor.Equals(other.Descriptor) ||
+            !Span.Equals(other.Span) ||
+            Args.Length != other.Args.Length)
         {
-            throw new ArgumentNullException(nameof(descriptor));
+            return false;
         }
 
-        return new DefaultRazorDiagnostic(descriptor, span, EmptyArgs);
+        for (var i = 0; i < Args.Length; i++)
+        {
+            if (!Args[i].Equals(other.Args[i]))
+            {
+                return false;
+            }
+        }
+
+        return true;
     }
 
-    public static RazorDiagnostic Create(RazorDiagnosticDescriptor descriptor, SourceSpan span, params object[] args)
+    public override int GetHashCode()
     {
-        if (descriptor == null)
+        var hash = HashCodeCombiner.Start();
+        hash.Add(Descriptor.GetHashCode());
+        hash.Add(Span.GetHashCode());
+
+        for (var i = 0; i < Args.Length; i++)
         {
-            throw new ArgumentNullException(nameof(descriptor));
+            hash.Add(Args[i]);
         }
 
-        return new DefaultRazorDiagnostic(descriptor, span, args);
+        return hash;
     }
 
     public override string ToString()
     {
-        return ((IFormattable)this).ToString(null, null);
-    }
-
-    public override bool Equals(object obj)
-    {
-        var other = obj as RazorDiagnostic;
-        return other == null ? false : Equals(other);
+        return ((IFormattable)this).ToString(format: null, formatProvider: null);
     }
 
-    string IFormattable.ToString(string ignore, IFormatProvider formatProvider)
+    string IFormattable.ToString(string format, System.IFormatProvider formatProvider)
```

Comment:
nit: is `System.` in `System.IFormatProvider` necessary here?

Reply:
Thanks! I missed that and will fix it in my next PR based on this one.